File: step_13_binary_trees/13.3_hard/P013_morris_preorder_traversal/solution.py
"""
Problem: Morris Preorder Traversal
Difficulty: HARD | XP: 50

O(1) space preorder traversal using threaded binary tree (Morris algorithm).
No stack, no recursion — temporarily modifies tree then restores it.
"""

from collections import deque
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1: BRUTE FORCE
# Iterative preorder using explicit stack
# Time: O(N)  |  Space: O(N)
# ============================================================
def brute_force(root: Optional[TreeNode]) -> List[int]:
    result = []
    if not root:
        return result
    stack = [root]
    while stack:
        node = stack.pop()
        result.append(node.val)
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)
    return result


# ============================================================
# APPROACH 2: OPTIMAL
# Morris Traversal — O(1) space threaded tree
# For each node: no left child => visit + go right.
# Has left child => find predecessor:
#   if pred.right is None => thread it, visit current, go left
#   if pred.right == current => unthread, go right
# Time: O(N)  |  Space: O(1)
# ============================================================
def optimal(root: Optional[TreeNode]) -> List[int]:
    result = []
    curr = root
    while curr:
        if not curr.left:
            result.append(curr.val)
            curr = curr.right
        else:
            # Find inorder predecessor
            pred = curr.left
            while pred.right and pred.right is not curr:
                pred = pred.right
            if pred.right is None:
                # First visit: thread and visit current (preorder)
                result.append(curr.val)
                pred.right = curr
                curr = curr.left
            else:
                # Second visit: unthread, move right
                pred.right = None
                curr = curr.right
    return result


# ============================================================
# APPROACH 3: BEST
# Recursive DFS — clean reference implementation
# Time: O(N)  |  Space: O(H) recursion stack
# ============================================================
def best(root: Optional[TreeNode]) -> List[int]:
    result = []

    def dfs(node):
        if not node:
            return
        result.append(node.val)
        dfs(node.left)
        dfs(node.right)

    dfs(root)
    return result


def build_tree(vals):
    if not vals:
        return None
    root = TreeNode(vals[0])
    queue = deque([root])
    i = 1
    while queue and i < len(vals):
        node = queue.popleft()
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            queue.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            queue.append(node.right)
        i += 1
    return root


if __name__ == "__main__":
    print("=== Morris Preorder Traversal ===")

    # Tree: 1->2,3; 2->4,5 => preorder [1,2,4,5,3]
    vals = [1, 2, 3, 4, 5]
    print(f"BruteForce:       {brute_force(build_tree(vals))}")  # [1,2,4,5,3]
    print(f"Optimal (Morris): {optimal(build_tree(vals))}")       # [1,2,4,5,3]
    print(f"Best (Recursive): {best(build_tree(vals))}")          # [1,2,4,5,3]

    # Skewed left tree: 1->2->3
    skewed = TreeNode(1, TreeNode(2, TreeNode(3)))
    print(f"Skewed Morris: {optimal(skewed)}")  # [1,2,3]

    # Single node
    print(f"Single node: {optimal(TreeNode(42))}")  # [42]

File: step_13_binary_trees/13.3_hard/P013_morris_preorder_traversal/Solution.java
import java.util.*;

/**
 * Problem: Morris Preorder Traversal
 * Difficulty: HARD | XP: 50
 *
 * O(1) space preorder traversal using threaded binary tree (Morris algorithm).
 * No stack, no recursion — modifies tree temporarily then restores it.
 *
 * @author DSA_Nova
 */

class MorrisTreeNode {
    int val;
    MorrisTreeNode left, right;
    MorrisTreeNode(int val) { this.val = val; }
}

// ============================================================
// APPROACH 1: BRUTE FORCE
// Iterative preorder using explicit stack
// Time: O(N)  |  Space: O(N)
// ============================================================
class BruteForce {
    static List<Integer> preorder(MorrisTreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        Deque<MorrisTreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            MorrisTreeNode node = stack.pop();
            result.add(node.val);
            if (node.right != null) stack.push(node.right);
            if (node.left != null) stack.push(node.left);
        }
        return result;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL
// Morris Traversal — standard implementation
// For each node: if no left child, visit and go right.
// Else, find inorder predecessor. If its right is null, thread it back
// to current (visit current now for preorder), go left.
// If threaded, unthread and go right.
// Time: O(N)  |  Space: O(1)
// ============================================================
class Optimal {
    static List<Integer> preorder(MorrisTreeNode root) {
        List<Integer> result = new ArrayList<>();
        MorrisTreeNode curr = root;
        while (curr != null) {
            if (curr.left == null) {
                result.add(curr.val);
                curr = curr.right;
            } else {
                // Find inorder predecessor
                MorrisTreeNode pred = curr.left;
                while (pred.right != null && pred.right != curr) {
                    pred = pred.right;
                }
                if (pred.right == null) {
                    // First visit: thread and visit current (preorder = visit before going left)
                    result.add(curr.val);
                    pred.right = curr;
                    curr = curr.left;
                } else {
                    // Second visit: unthread, move right
                    pred.right = null;
                    curr = curr.right;
                }
            }
        }
        return result;
    }
}

// ============================================================
// APPROACH 3: BEST
// Recursive preorder — simplest correct reference implementation
// Time: O(N)  |  Space: O(H) stack
// ============================================================
class Best {
    static List<Integer> preorder(MorrisTreeNode root) {
        List<Integer> result = new ArrayList<>();
        dfs(root, result);
        return result;
    }

    static void dfs(MorrisTreeNode node, List<Integer> result) {
        if (node == null) return;
        result.add(node.val);
        dfs(node.left, result);
        dfs(node.right, result);
    }
}

public class Solution {
    static MorrisTreeNode build(Integer[] vals) {
        if (vals == null || vals.length == 0) return null;
        MorrisTreeNode root = new MorrisTreeNode(vals[0]);
        Queue<MorrisTreeNode> q = new LinkedList<>();
        q.offer(root);
        int i = 1;
        while (!q.isEmpty() && i < vals.length) {
            MorrisTreeNode cur = q.poll();
            if (i < vals.length && vals[i] != null) { cur.left = new MorrisTreeNode(vals[i]); q.offer(cur.left); }
            i++;
            if (i < vals.length && vals[i] != null) { cur.right = new MorrisTreeNode(vals[i]); q.offer(cur.right); }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        System.out.println("=== Morris Preorder Traversal ===");

        // Tree: 1->2,3; 2->4,5
        Integer[] vals = {1, 2, 3, 4, 5};
        // Expected preorder: [1, 2, 4, 5, 3]

        System.out.println("BruteForce: " + BruteForce.preorder(build(vals)));
        System.out.println("Optimal (Morris): " + Optimal.preorder(build(vals)));
        System.out.println("Best (Recursive): " + Best.preorder(build(vals)));

        // Skewed left tree
        MorrisTreeNode skewed = new MorrisTreeNode(1);
        skewed.left = new MorrisTreeNode(2);
        skewed.left.left = new MorrisTreeNode(3);
        System.out.println("Skewed Morris: " + Optimal.preorder(skewed)); // [1,2,3]

        // Single node
        System.out.println("Single node: " + Optimal.preorder(new MorrisTreeNode(42))); // [42]
    }
}
